src/python/deepgnn/graph_engine/snark/decoders.py

```diff
@@ -43,6 +43,199 @@ def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        * Feature vectors are given indicies based on the order they appear in the line, the first feature vector is index 0.
+        A feature index can be skipped by giving a 0 length vector.
+
+    Edge List Format Example
+        A graph with 2 nodes {0, 1} each with type = 1, weight = .5 and
+        feature vectors [1, 1, 1] dtype=int32 and [1.1, 1.1] dtype=float32.
+        Edges: {0 -> 1, 1 -> 0} both with type = 0, weight = .5 and a sparse feature
+        vector (coords=[0, 4, 10], values=[1, 1, 1] dtype=uint8).
+        ```
+        0,-1,1,.5,int32,3,1,1,1,float32,2,1.1,1.1
+        0,1,0,.5,uint8,3/0,0,4,10,1,1,1
+        1,-1,1,.5,int32,3,1,1,1,float32,2,1.1,1.1
+        1,0,0,.5,uint8,3/0,0,4,10,1,1,1
+        ```
+
+    Init Parameters:
+        the following keyword arguments can be added when creating the decoder.
+        default_node_type: int Type of all nodes, if set do not add node type to any nodes.
+        default_node_weight: int Weight of all nodes, if set do not add node weight to any nodes.
+        default_node_feature_types: ["dtype" or None, ...] Dtype of each feature vector.
+        default_node_feature_lens: [[int, ...] or None, ...] Length value for each feature vector.
+        default_edge_type: int Same as node except for all edges.
+        default_edge_weight: int Same as node except for all edges.
+        default_edge_feature_types: ["dtype" or None, ...] Dtype of each feature vector.
+        default_edge_feature_lens: [[int, ...] or None, ...] Length value for each feature vector.
+
+    Init Parameters Format Example
+        Same graph as the previous example, just with defaults specified.
+        ```
+        EdgeListDecoder(
+            default_node_type=1,
+            default_node_weight=.5,
+            default_node_feature_types=["int32", "float32"],
+            default_node_feature_lens=[[3],[2]],
+            default_edge_type=0,
+            default_edge_weight=.5,
+            default_edge_feature_types=["uint8"],
+                default_edge_feature_lens=[[3, 0]],
+        )
+        ```
+        graph.csv
+        ```
+        0,-1,1,1,1,1.1,1.1
+        0,1,0,4,10,1,1,1
+        1,-1,1,1,1,1.1,1.1
+        1,0,0,4,10,1,1,1
+        ```
+
+
+    * \ is the escape for the delimiter in "binary" features.
+    """
+
+    def __init__(
+        self,
+        default_node_type: Optional[int] = None,
+        default_edge_type: Optional[int] = None,
+        default_node_weight: Optional[float] = None,
+        default_edge_weight: Optional[float] = None,
+        default_node_feature_types: Optional[List[Optional[str]]] = None,
+        default_edge_feature_types: Optional[List[Optional[str]]] = None,
+        default_node_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        default_edge_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        delimiter=",",
+        length_delimiter="/",
```

Comment:
how do you escape these delimiters for string features?

Reply:
Its with a backslash, Its at the bottom of the docstring.